File: aimp_dotnet/Extensions/AimpExtensionAlbumArtProvider.cpp
#include "Stdafx.h"
#include "AimpExtensionAlbumArtProvider.h"
#include "..\SDK\PlayList\AimpFileInfo.h"
#include "..\SDK\AlbumArt\AimpAlbumArtSearchOptions.h"

HRESULT WINAPI AimpExtensionAlbumArtProvider::Get(IAIMPString *FileURI, IAIMPString *Artist, IAIMPString *Album, IAIMPPropertyList *Options, IAIMPImageContainer **Image)
{
    System::Drawing::Bitmap ^bitmap;
    AIMP::SDK::AlbumArtManager::IAimpAlbumArtSearchOptions ^searchOptions = gcnew AIMP::SDK::AimpAlbumArtSearchOptions(Options, AIMP::SDK::ManagedAimpCore::GetAimpCore());
    AIMP::AimpActionResult r = _managedinstance->Get(
        gcnew System::String(FileURI->GetData()),
        gcnew System::String(Artist->GetData()),
        gcnew System::String(Album->GetData()),
        searchOptions,
        bitmap);
    {
        IAIMPImageContainer *container = AIMP::SDK::AimpExtension::ToContainer(bitmap);
        if (container == NULL)
        {
            return E_UNEXPECTED;
        }

        *Image = container;
        return S_OK;
    }

    return E_UNEXPECTED;
}

DWORD WINAPI AimpExtensionAlbumArtProvider::GetCategory()
{
    return (DWORD)_managedinstance->GetCategory();
}

HRESULT WINAPI AimpExtensionAlbumArtProvider::Get2(IAIMPFileInfo *FileInfo, IAIMPPropertyList *Options, IAIMPImageContainer **Image)
{
    System::Drawing::Bitmap^ bitmap;
    AIMP::SDK::FileManager::IAimpFileInfo ^fi = gcnew AIMP::SDK::AimpFileInfo(FileInfo);
    AIMP::SDK::AlbumArtManager::IAimpAlbumArtSearchOptions ^searchOptions = gcnew AIMP::SDK::AimpAlbumArtSearchOptions(Options, AIMP::SDK::ManagedAimpCore::GetAimpCore());
    AIMP::AimpActionResult r = _managedinstance->Get(fi, searchOptions, bitmap);
    if (r == AIMP::AimpActionResult::Ok && bitmap != nullptr)
    {
        IAIMPImageContainer *container = AIMP::SDK::AimpExtension::ToContainer(bitmap);
        if (container == NULL)
        {
            return E_UNEXPECTED;
        }

        *Image = container;
        return S_OK;
    }

    return E_UNEXPECTED;
}

File: Plugins/dotnet_albumart/AimpExtensionAlbumArtProvider.cs
namespace dotnet_albumart
{
    using System.Drawing;
    using System.Runtime.InteropServices.WindowsRuntime;

    using AIMP.SDK;
    using AIMP.SDK.AlbumArtManager;
    using AIMP.SDK.PlayList;

    public class AimpExtensionAlbumArtProvider : IAimpExtensionAlbumArtProvider
    {
        #region Implementation of IAimpExtensionAlbumArtProvider2

        public AimpActionResult Get(IAimpFileInfo fileInfo, IAimpAlbumArtSearchOptions options, out Bitmap image)
        {
            image = Properties.Resources.front;
            return AimpActionResult.Ok;
        }

        #endregion

        #region Implementation of IAimpExtensionAlbumArtProvider

        public AimpActionResult Get(string fileUrl, string artist, string album, IAimpAlbumArtSearchOptions options, out Bitmap image)
        {
            image = Properties.Resources.front;
            return AimpActionResult.Ok;
        }

        #endregion

        #region Implementation of IAimpExtensionAlbumArtProvider

        public AimpAlbumArtProviderCategory GetCategory()
        {
            return AimpAlbumArtProviderCategory.AIMP_ALBUMART_PROVIDER_CATEGORY_INTERNET;
        }

        #endregion
    }
}